### ISSUE TITLE
System.TypeLoadException when returning this from an abstract method with generic types

### ISSUE THREAD

**Issue by cesarsouza** [dotnet/roslyn#50649]
It seems either Roslyn does not detect an invalid program during compilation or there is a problem with the runtime. I am sending a repro project attached demonstrating the issue. It may be that the program is stepping in undefined behavior, but the compiler does not produce any warnings if that is the case.

The reproduction project contains two main entrypoints: one which generates a TypeLoadException (see below) and another which works. The only difference between the two is that, in the example that produces the exception, all types have a generic parameter (called T1). In the example that works, the involved types do not have T1 as the first generic parameter.

**Version Used**: 

Microsoft Visual Studio Community 2019 Preview
Version 16.9.0 Preview 1.0
VisualStudio.16.Preview/16.9.0-pre.1.0+30709.64
Microsoft .NET Framework
Version 4.8.03752

**Steps to Reproduce**:

1.  Download the repro example
2.  Hit F5, should get exception below
3.  Change entrypoint to "Program_Works.Main"
4.  Hit F5, should work

**Expected Behavior**:

If the program is not valid, Roslyn should at least provide a warning about it.

**Actual Behavior**:

Program compiles with no errors but fails at runtime with the following exception:

System.TypeLoadException
  HResult=0x80131522
  Message=Return type in method 'Repro.Program_TypeLoadException+Concrete`1[T1].get_This()' on type 'Repro.Program_TypeLoadException+Concrete`1[T1]' from assembly 'Repro, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not compatible with base type method 'Repro.Program_TypeLoadException+Concrete`2[T1,System.__Canon].get_This()'.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

[repro-typeload.zip](https://github.com/dotnet/roslyn/files/5844860/repro-typeload.zip)

**Comment by jaredpar** [comment]
@AlekseyTs the issue is around covariant returns. Unsure at a glance of the IL if this is our side or runtime side.

**Comment by AlekseyTs** [comment]
I believe this is an issue in runtime. I can reproduce the problem with smaller scenario:
```
public static class Program_TypeLoadException
{
    public static void Main(string[] args)
    {
        var node = new D<object>();
    }
}

public abstract class C<CT1, CT2>
{
    public C()
    {
        this.This = this;
    }

    protected virtual C<CT1, CT2> This { get; }
}

public class D<DT1> : C<DT1, object>
{
    public D()
    {
        this.This = this;
    }

    protected override D<DT1> This { get; }
}
```

Observed:
```
Unhandled exception. System.TypeLoadException: 
Return type in method 'D`1[DT1].get_This()' on type 'D`1[DT1]' from assembly 'ConsoleAppIssue50649, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' 
is not compatible with base type method 'C`2[DT1,System.__Canon].get_This()'.
   at Program_TypeLoadException.Main(String[] args) 
```

As long as base class of ```D``` refers to ```DT1``` in any position, there is a type load exception. If base class is changed to ```C<object, object>```, there is no exception.

**Comment by janvorli** [comment]
@cesarsouza this was already fixed by https://github.com/dotnet/runtime/pull/47048, which should be released in 5.0.3 (porting PR approved https://github.com/dotnet/runtime/pull/47233).
